lib/kdl/include/kdl/compact_trie.h

```diff
@@ -0,0 +1,714 @@
+             *
+             * Precondition: This node's key has at least two characters, and the index is chosen in such a way that
+             * neither of the resulting keys is empty.
+             *
+             * @param index the index at which to split the node's key
+             */
+            void split_node(const std::size_t index) const {
+                assert(m_key.length() > 1u);
+
+                auto new_key = m_key.substr(0u, index);
+                auto remainder = m_key.substr(index);
+
+                assert(!new_key.empty());
+                assert(!remainder.empty());
+
+                using std::swap;
+                node_set new_children;
+                swap(new_children, m_children);
+
+                const node& new_child = *m_children.insert(node(std::move(remainder))).first;
+                swap(new_child.m_children, new_children);
+                swap(new_child.m_values, m_values);
+
+                m_key = std::move(new_key);
+            }
+
+            /**
+             * Merges this node with its only child. Thereby, this child node's key is appended to this node's key, the
+             * child's children and values are moved to this node, and the child is removed.
+             *
+             * Precondition: This node has only one child, and this node has no values of its own.
+             */
+            void merge_node() const {
+                assert(m_children.size() == 1u);
+                assert(m_values.empty());
+
+                using std::swap;
+                node_set old_children;
+                swap(old_children, m_children);
+
+                const node& child = *std::begin(old_children);
+                swap(m_children, child.m_children);
+                swap(m_values, child.m_values);
+
+                m_key += child.m_key;
+            }
+
+            template <typename O>
+            void get_values(O out) const {
+                for (const auto& [value, count] : m_values) {
+                    for (std::size_t i = 0u; i < count; ++i) {
+                        out++ = value;
+                    }
+                }
+            }
+
+            template <typename O>
+            void get_values_and_recurse(O out) const {
+                get_values(out);
+                for (const auto& child : m_children) {
+                    child.get_values_and_recurse(out);
+                }
+            }
+        };
+
+        /**
+         * Compares nodes against each other or nodes against strings.
+         *
+         * Two nodes are compared by their keys.
+         * A node is compared to a string by comparing its key to the string.
+         *
+         * If two nodes share a non-empty prefix, they are considered equivalent. Therefore, the keys "ab" and "abc" are
+         * equivalent. Conversely, the keys "ab" and "bc" are not equivalent because they do not share a prefix, more
+         * specifically, "ab" is considered less than "bc".
```

Comment:
This is a roundabout description.. What this class actually does is compare the first character only

Reply:
I have extended the description.